badger/badger.go

```diff
@@ -0,0 +1,307 @@
+package badger
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/dgraph-io/badger"
+	"github.com/pkg/errors"
+	nosql "github.com/smallstep/nosql/database"
+)
+
+// DB is a wrapper over *badger.DB,
+type DB struct {
+	db *badger.DB
+}
+
+// Open opens or creates a BoltDB database in the given path.
+func (db *DB) Open(dir string, opt nosql.Options) (err error) {
+	bo := badger.DefaultOptions
+	bo.Dir = dir
+	if opt.ValueDir != "" {
+		bo.ValueDir = opt.ValueDir
+	} else {
+		bo.ValueDir = dir
+	}
+
+	db.db, err = badger.Open(bo)
+	return errors.Wrap(err, "error opening Badger database")
+}
+
+// Close closes the DB database.
+func (db *DB) Close() error {
+	return errors.Wrap(db.db.Close(), "error closing Badger database")
+}
+
+// CreateTable is a noop because badger does not support tables, buckets,
+// collections, etc.
+func (db *DB) CreateTable(bucket []byte) error {
+	return nil
+}
+
+// DeleteTable deletes a root or embedded bucket. Returns an error if the
+// bucket cannot be found or if the key represents a non-bucket value.
+func (db *DB) DeleteTable(bucket []byte) error {
+	prefix, err := badgerEncode(bucket)
+	if err != nil {
+		return err
+	}
+	deleteKeys := func(keysForDelete [][]byte) error {
+		if err := db.db.Update(func(txn *badger.Txn) error {
+			for _, key := range keysForDelete {
+				if err := txn.Delete(key); err != nil {
+					return errors.Wrapf(err, "error deleting key %s", key)
+				}
+			}
+			return nil
+		}); err != nil {
+			return errors.Wrapf(err, "update failed")
+		}
+		return nil
+	}
+
+	collectSize := 1000
+	err = db.db.View(func(txn *badger.Txn) error {
+		opts := badger.DefaultIteratorOptions
+		opts.AllVersions = false
+		opts.PrefetchValues = false
+		it := txn.NewIterator(opts)
+		defer it.Close()
+
+		keysForDelete := make([][]byte, 0, collectSize)
+		keysCollected := 0
+		for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
+			key := it.Item().KeyCopy(nil)
+			keysForDelete = append(keysForDelete, key)
+			keysCollected++
+			if keysCollected == collectSize {
+				if err := deleteKeys(keysForDelete); err != nil {
+					return err
+				}
+				keysForDelete = make([][]byte, 0, collectSize)
+				keysCollected = 0
+			}
+		}
+		if keysCollected > 0 {
+			if err := deleteKeys(keysForDelete); err != nil {
+				return err
+			}
+		}
```

Comment:
Use always a fixed sized array instead of creating one every 1000 items and send to `deleteKeys` just the portion filled `deleteKeys(keysForDelete[:keysCollected]`